File: backend/app/domain/provider.py
"""
Provider Domain Model

Defines Provider related Data Transfer Objects (DTOs).
"""

import logging
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.common.provider_protocols import FRONTEND_PROTOCOL_PATTERN
from app.common.url_validator import validate_provider_url_loose

logger = logging.getLogger(__name__)

DEFAULT_RESPONSE_TIMEOUT_SECONDS = 1800


class ProviderBase(BaseModel):
    """Provider Base Model"""

    # Provider Name
    name: str = Field(..., min_length=1, max_length=100, description="Provider Name")
    # Remark
    remark: Optional[str] = Field(None, max_length=500, description="Provider Remark")
    # Base URL
    base_url: str = Field(..., description="Base URL")
    # Protocol Type (frontend protocol)
    protocol: str = Field(..., pattern=FRONTEND_PROTOCOL_PATTERN, description="Protocol Type")
    # API Type: chat / completion / embedding
    api_type: str = Field("chat", description="API Type (deprecated)")
    # Extra Headers
    extra_headers: Optional[dict[str, str]] = Field(None, description="Extra Headers")
    # Provider Options (JSON format)
    provider_options: Optional[dict[str, Any]] = Field(
        None, description="Provider Options"
    )
    # Proxy Enabled
    proxy_enabled: bool = Field(False, description="Proxy Enabled")
    # Proxy URL (schema://auth@host:port)
    proxy_url: Optional[str] = Field(None, description="Proxy URL")
    # No-response timeout in seconds
    response_timeout_seconds: int = Field(
        DEFAULT_RESPONSE_TIMEOUT_SECONDS,
        ge=1,
        description="No-response timeout in seconds",
    )

    @field_validator("base_url")
    @classmethod
    def validate_base_url(cls, v: str) -> str:
        """Validate base_url to prevent SSRF"""
        try:
            return validate_provider_url_loose(v)
        except Exception as e:
            logger.warning("Provider base_url validation failed: %s", str(e))
            raise ValueError(f"Invalid base_url: {str(e)}")


class ProviderCreate(ProviderBase):
    """Create Provider Request Model"""
    
    # Provider API Key (Optional)
    api_key: Optional[str] = Field(None, description="Provider API Key")
    # Is Active
    is_active: bool = Field(True, description="Is Active")


class ProviderUpdate(BaseModel):
    """Update Provider Request Model (All fields optional)"""
    
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    remark: Optional[str] = Field(None, max_length=500)
    base_url: Optional[str] = None
    protocol: Optional[str] = Field(None, pattern=FRONTEND_PROTOCOL_PATTERN)
    api_type: Optional[str] = None
    api_key: Optional[str] = None
    extra_headers: Optional[dict[str, str]] = None
    provider_options: Optional[dict[str, Any]] = None
    is_active: Optional[bool] = None
    proxy_enabled: Optional[bool] = None
    proxy_url: Optional[str] = None
    response_timeout_seconds: Optional[int] = Field(None, ge=1)


class Provider(ProviderBase):
    """Provider Complete Model"""
    
    id: int = Field(..., description="Provider ID")
    api_key: Optional[str] = Field(None, description="Provider API Key")
    extra_headers: Optional[dict[str, str]] = Field(None, description="Extra Headers")
    provider_options: Optional[dict[str, Any]] = Field(
        None, description="Provider Options"
    )
    is_active: bool = Field(True, description="Is Active")
    created_at: datetime = Field(..., description="Creation Time")
    updated_at: datetime = Field(..., description="Update Time")
    
    model_config = ConfigDict(from_attributes=True)


class ProviderResponse(ProviderBase):
    """Provider Response Model (API Key Sanitized)"""
    
    id: int = Field(..., description="Provider ID")
    # API Key Sanitized Display
    api_key: Optional[str] = Field(None, description="Provider API Key (Sanitized)")
    extra_headers: Optional[dict[str, str]] = Field(None, description="Extra Headers")
    provider_options: Optional[dict[str, Any]] = Field(
        None, description="Provider Options"
    )
    proxy_url: Optional[str] = Field(None, description="Proxy URL (Sanitized)")
    is_active: bool = Field(True, description="Is Active")
    created_at: datetime = Field(..., description="Creation Time")
    updated_at: datetime = Field(..., description="Update Time")
    
    model_config = ConfigDict(from_attributes=True)


class ProviderNameResponse(BaseModel):
    """Provider name list item for selector APIs."""

    id: int = Field(..., description="Provider ID")
    name: str = Field(..., description="Provider Name")
    protocol: str = Field(..., description="Protocol Type")
    is_active: bool = Field(True, description="Is Active")

    model_config = ConfigDict(from_attributes=True)


class ProviderExport(ProviderCreate):
    """Provider Export Model (Includes API Key)"""
    pass

File: frontend/src/types/provider.ts
/**
 * Provider Related Type Definitions
 * Corresponds to backend service_providers table
 */

/** Protocol Type */
export type ImplementationProtocolType = string;
export type ProtocolType = string;

export interface ProviderOptions {
  default_parameters?: Record<string, unknown>;
  no_suffix?: boolean;
}

/** Provider Entity */
export interface Provider {
  id: number;
  name: string;
  remark?: string;
  base_url: string;
  protocol: ProtocolType;
  api_key?: string;          // Sanitized display
  extra_headers?: Record<string, string>;
  provider_options?: ProviderOptions;
  proxy_enabled?: boolean;
  proxy_url?: string; // Sanitized display
  response_timeout_seconds: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

/** Provider name list item */
export interface ProviderName {
  id: number;
  name: string;
  protocol: ProtocolType;
  is_active: boolean;
}

/** Create Provider Request */
export interface ProviderCreate {
  name: string;
  remark?: string;
  base_url: string;
  protocol: ProtocolType;
  api_key?: string;
  extra_headers?: Record<string, string>;
  provider_options?: ProviderOptions;
  proxy_enabled?: boolean;
  proxy_url?: string;
  response_timeout_seconds?: number;
  is_active?: boolean;
}

/** Update Provider Request */
export interface ProviderUpdate {
  name?: string;
  remark?: string;
  base_url?: string;
  protocol?: ProtocolType;
  api_key?: string;
  extra_headers?: Record<string, string>;
  provider_options?: ProviderOptions;
  proxy_enabled?: boolean;
  proxy_url?: string;
  response_timeout_seconds?: number;
  is_active?: boolean;
}

/** Provider List Query Params */
export interface ProviderListParams {
  is_active?: boolean;
  page?: number;
  page_size?: number;
  name?: string;
  protocol?: ProtocolType;
}

export interface ProviderProtocolConfig {
  protocol: ProtocolType;
  label: string;
  implementation: ImplementationProtocolType;
  base_url: string;
}

/** Provider Model List Response */
export interface ProviderModelListResponse {
  provider_id: number;
  provider_name: string;
  protocol: ProtocolType;
  models: string[];
  success: boolean;
  error?: {
    message: string;
    code: string;
    details?: Record<string, unknown>;
  };
}

/** Provider Export Entity (includes API key) */
export type ProviderExport = ProviderCreate;
